### PR TITLE
CO-2290 Sync keys without transaction

### PR DESCRIPTION
**Pull request by willhr3**
- Added a constructor to the SyncKeyMappings flow to allow a node to generate mappings for a list of CI directly rather than having to extract them from a transaction.

### CODE REVIEW

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt [hunk 1]
```diff
@@ -14,14 +14,21 @@ import net.corda.core.utilities.unwrap
...
+ * will be extracted and then registered in the identity service. Alternatively, the node can provide the confidential identities
+ * it wishes to share with the counter-party directly. The initiating node sends the list of identities to the counter-party who
+ * registers the identity mapping between a newly generated [PublicKey] and the [Party].
  *
  * The counter-party will request a new key mapping for each of the unresolved identities by calling [RequestKeyFlow] as
  * an inline flow.
  */
-class SyncKeyMappingFlow(private val session: FlowSession, private val tx: WireTransaction) : FlowLogic<Unit>() {
+class SyncKeyMappingFlow(
+        private val session: FlowSession,
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: Same constructor should be private in this case.

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt [hunk 2]
```diff
@@ -14,14 +14,21 @@ import net.corda.core.utilities.unwrap
 import java.security.PublicKey
 
 /**
- * This flow allows a node to share the [PublicKey] to [Party] mapping data of parties unknown parties present in a given
- * transaction. The initiating node sends a list of confidential identities to the counter-party who attempts to resolve
- * them. Parties that cannot be resolved are returned to the initiating node.
+ * This flow allows a node to share the [PublicKey] to [Party] mapping data of parties unknown to the counter-party. The
+ * two constructors allow the the node to provide a transaction in which the confidential identities unknown to the counter-party
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: typo: the the node

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt [hunk 3]
```diff
@@ -33,12 +40,22 @@ class SyncKeyMappingFlow(private val session: FlowSession, private val tx: WireT
...
+            identitiesToSync?.forEach {
+                confidentialIdentities.add(it)
+            }
+        }
 
         // Send confidential identities to the counter party and return a list of parties they wish to resolve
         val requestedIdentities = session.sendAndReceive<List<AbstractParty>>(confidentialIdentities).unwrap { req ->
             require(req.all { it in confidentialIdentities }) {
-                "${session.counterparty} requested a confidential identity not part of transaction: ${tx.id}"
+                "${session.counterparty} requested a confidential identity not part of transaction: ${tx?.id}"
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: This isn't very useful error it tx is null. Maybe change it.

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt [hunk 4]
```diff
@@ -48,6 +65,8 @@ class SyncKeyMappingFlow(private val session: FlowSession, private val tx: WireT
     }
 
     private fun extractConfidentialIdentities(): List<AbstractParty> {
+        tx
+                ?: throw IllegalArgumentException("A transaction must be provided if you wish to extract the confidential identities from it.")
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: put it in one line

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt [hunk 5]
```diff
@@ -48,6 +65,8 @@ class SyncKeyMappingFlow(private val session: FlowSession, private val tx: WireT
     }
 
     private fun extractConfidentialIdentities(): List<AbstractParty> {
+        tx
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: even better make function take as a parameter transaction, then you don't need to do additional nullability check (you already did it in call site)

**Comment by willhr3** [comment]
Hi @kasiastreich this PR is ready for a second round of review - cheers.

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/FlowUtilities.kt [hunk 6]
```diff
@@ -71,17 +73,27 @@ class ShareKeyInitiator(private val otherParty: Party, private val uuid: UUID) :
...
+        private val otherParty: Party,
+        private val tx: WireTransaction?,
+        private val identitiesToSync: List<AbstractParty>?) : FlowLogic<Unit>() {
+    constructor(otherParty: Party, tx: WireTransaction) : this(otherParty, tx, null)
+    constructor(otherParty: Party, identitiesToSync: List<AbstractParty>) : this(otherParty, null, identitiesToSync)
+
     @Suspendable
     override fun call() {
-        subFlow(SyncKeyMappingFlow(initiateFlow(otherParty), tx))
+        when {
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/FlowUtilities.kt]: use if else

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt (hunk 2 above)
Comment (kasiastreich) [reply workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: Still unresolved. Please don't mark comments as resolved when you didn't address them

File: workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt [hunk 7]
```diff
@@ -33,12 +41,22 @@ class SyncKeyMappingFlow(private val session: FlowSession, private val tx: WireT
     @Suspendable
     override fun call() {
         progressTracker.currentStep = SYNCING_KEY_MAPPINGS
-        val confidentialIdentities = extractConfidentialIdentities()
+        val confidentialIdentities = mutableListOf<AbstractParty>()
+        if (tx != null) {
+            val ci = extractConfidentialIdentities(tx)
```
Comment (kasiastreich) [inline workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: You can add directly to `confidentialIdentities` without `ci` variable.

Comment (kasiastreich) [reply workflows/src/main/kotlin/com/r3/corda/lib/ci/SyncKeyMappingFlow.kt]: Or even better, you can write:
`val confidentialIdentities = if().. else()...`

**Comment by GavT** [comment]
@willhr3 I updated the title to correct the Jira ref from CO-2290 to CORDA-2290

**Pull request merged**